### PR TITLE
Add faster implementation of _IonNature.ion_hash

### PR DESCRIPTION
**Pull request by popematt**
*Issue #, if available:*

Not necessarily a fix; at least an improvement for #21 

*Description of changes:*

New implementation of `.ion_hash()` that does away with the overhead of instantiating `ion_writer`s. As part of this change, I  improved the performance of the `_escape()` function, which has the side effect of also benefitting the `ion_hash_reader` and the `ion_hash_writer`.

I did some _very_ informal performance testing using [this script](https://gist.github.com/popematt/a847ac825ffb267f115599cc0ec983f9).

```
                      Time to hash single value             Ops/Sec                   Approx
Test Case                 (old)       (new)           (old)             (new)       Improvement
===============================================================================================
ion_hash_python_21     2734.39 μs   389.30 μs      365.71 ops/s      2568.72 ops/s       7x
int                     106.84 μs     9.36 μs     9359.53 ops/s    106820.70 ops/s      11x
bool                    100.58 μs     6.17 μs     9942.43 ops/s    162009.17 ops/s      16x
null                     98.02 μs     7.11 μs    10201.86 ops/s    140577.22 ops/s      14x
string                  104.39 μs     7.59 μs     9579.21 ops/s    131760.86 ops/s      14x
decimal                 125.09 μs    16.29 μs     7994.50 ops/s     61396.71 ops/s       8x
clob                    102.30 μs     8.38 μs     9775.37 ops/s    119359.82 ops/s      12x
timestamp               178.36 μs    40.24 μs     5606.68 ops/s     24850.82 ops/s       4x
empty_struct            139.38 μs     7.16 μs     7174.65 ops/s    139569.14 ops/s      19x
struct_one_field        307.97 μs    38.24 μs     3247.12 ops/s     26150.63 ops/s       8x
struct_few_fields       572.92 μs    75.29 μs     1745.44 ops/s     13281.10 ops/s       8x
struct_many_fields     1735.05 μs   329.23 μs      576.35 ops/s      3037.42 ops/s       5x
empty_list              118.40 μs     6.21 μs     8445.82 ops/s    161004.19 ops/s      19x
long_list              2890.12 μs   180.54 μs      346.01 ops/s      5539.09 ops/s      16x
nested_containers       487.38 μs    19.84 μs     2051.80 ops/s     50411.10 ops/s      25x
one_annotation          265.96 μs    11.36 μs     3760.01 ops/s     87993.25 ops/s      23x
many_annotations        506.17 μs    20.03 μs     1975.62 ops/s     49921.85 ops/s      25x
very_long_string        246.75 μs    10.28 μs     4052.64 ops/s     97308.19 ops/s      24x
very_long_blob          304.76 μs    15.67 μs     3281.30 ops/s     63807.71 ops/s      19x

```

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Comment by cheqianh** [comment]
Look pretty good!

QQ, did the change pass all the unit test (e.g. pytest under the `tests/`)? Just in case I missed any corner case.

**Review by cheqianh — COMMENTED** [review]
Looks pretty good!

File: ionhash/__init__.py [hunk 1]
```diff
@@ -11,14 +11,9 @@
...
-from amazon.ion.core import ION_STREAM_END_EVENT
 from amazon.ion.simple_types import _IonNature
-from amazon.ion.simpleion import _dump, _FROM_TYPE
-from amazon.ion.writer import blocking_writer
-from amazon.ion.writer_binary import binary_writer
 
+from ionhash.fast_value_hasher import hash_value
 from ionhash.hasher import hashlib_hash_function_provider
 from ionhash.hasher import hash_writer
 from ionhash.hasher import HashEvent
```
Comment (tgregg) [inline ionhash/__init__.py]: Do we need either of these imports?

File: ionhash/hasher.py [hunk 2]
```diff
@@ -475,19 +475,12 @@ def _bytearray_comparator(a, b):
...
-            for c in _bytes:
-                if c == _BEGIN_MARKER_BYTE or c == _END_MARKER_BYTE or c == _ESCAPE_BYTE:
-                    escaped_bytes.append(_ESCAPE_BYTE)
-                escaped_bytes.append(c)
-            return escaped_bytes
-
+    if _BEGIN_MARKER_BYTE in _bytes or _END_MARKER_BYTE in _bytes or _ESCAPE_BYTE in _bytes:
+        return _bytes.replace(bytes([_ESCAPE_BYTE]), bytes([_ESCAPE_BYTE, _ESCAPE_BYTE])) \
+                     .replace(_BEGIN_MARKER, bytes([_ESCAPE_BYTE, _BEGIN_MARKER_BYTE])) \
+                     .replace(_END_MARKER, bytes([_ESCAPE_BYTE, _END_MARKER_BYTE]))
```
Comment (tgregg) [inline ionhash/hasher.py:483]: Should we avoid recomputing `bytes([...])` on each call by storing the values?

**Review by tgregg — APPROVED** [review]
This is very nice.

**Review by cheqianh — APPROVED** [review]

File: ionhash/__init__.py (hunk 1 above)
Comment (popematt) [reply ionhash/__init__.py]: Nope. Good catch.

File: ionhash/hasher.py (hunk 2 above)
Comment (popematt) [reply ionhash/hasher.py:483]: I don't know. I have also heard that looking up non-local values can be expensive in Python, and if we want to compute once and store the value, then it would need to be a non-local value. I'll create a followup issue to see if it would improve performance.

Comment (tgregg) [reply ionhash/hasher.py:483]: That's fine with me, although it does seem like a simple enough thing to try given that you already have a basic perf testing setup.
